[WJ-440] Client/frontend monorepo design (#181)
scpwiki/wikijump: client/docs/modules.md

# Monorepo: Modules

> Modules are the packages that can be found in the `modules/` folder. They're intended for pure ESM output, and are compiled rapidly using Estrella, an esbuild-based utility. They have an important file structure (don't change it), which looks like the following:

Reviewer: ```suggestion
Modules are the packages that can be found in the `modules/` folder. They're intended for pure ESM output, and are compiled rapidly using [Estrella](https://github.com/rsms/estrella), an [esbuild](https://github.com/evanw/esbuild)-based utility. They have an important file structure (don't change it), which looks like the following:
```

> Modules are the packages that can be found in the `modules/` folder. They're intended for pure ESM output, and are compiled rapidly using Estrella, an esbuild-based utility. They have an important file structure (don't change it), which looks like the following:

Reviewer: ```suggestion
Modules are the packages that can be found in the `modules/` folder. They're intended to produce self-contained ES modules for consumption by other utilities, and are compiled rapidly using Estrella, an esbuild-based utility. They have an important file structure (don't change it), which looks like the following:
```

```
module:
- src/
  - index.ts
  - **/*.ts
- tests/
  - *.ts
- package.json
- tsconfig.json
```

> The `src/` folder is compiled into a `dist/` folder when the module is built. Similarly, files in the `tests/` folder are compiled into the `tests/dist/` folder when the tests are built.

Reviewer 2: n/b: Is it `tests/dist/`, or does it go in `dist/tests/`?

Author: It goes into `tests/dist` because `dist/` is effectively the emitted content of the NPM package, which shouldn't have built test files in it.

> Other than the inheritance of development dependencies and tooling configurations, these modules act like fairly normal NPM packages. If you need to add a dependency to one, you can either navigate to the module and do a normal `npm i <dep>` command, or manually add it to the `package.json`. After you do that, you need to make sure to run the `npm run bootstrap` command at the root. This ensures that all the fragile symlinks get restored by Lerna.

Reviewer 2: Do you mean `npm i --save <dep>`?

Author: `npm i <dep>` automatically adds the package as a dependency. If it didn't, that would be kinda weird.

Reviewer 2: It does? That is not my understanding of how `npm install` works.

Okay then.

Author: Just to explain why it would be weird - NPM tries to keep your `node_modules` folder clean and deduped, and minimizes anything that _isn't_ your dependencies. So, if you install a package, didn't save it, then installed another one, the first one would get deleted. So, it's assumed installing a package means adding it as a dependency, which it is 99% of the time.

> To improve performance of the monorepo, building and watching of modules isn't done per module, but instead globally at root. It is recommended that you just build the entire monorepo rather than trying to build individual packages. This is not an insane recommendation because module building is 100% incremental. If a module has not changed, it won't get rebuilt.

Reviewer 2: "insane recommendation" is a weird way of saying it, say something like "This is not a performance issue because ..."

There is a template for making modules, and it can be found in the `templates/module-template` folder.

Here is how to use that template:

1. Create a folder in `modules/` that is the name of your package.

>   It isn't strictly required that it is the name of your package, but it gets confusing otherwise.

Reviewer 2: This seems more like a footnote. I don't think we should ever have a case where the name and directory mismatch, the documentation should just assert that they're the same.

Author: Well, we have that case already when you prefix with `node-`. I feel like that information is important because I don't want to hide how the repo works fundamentally.

Reviewer: I think I agree with ais here - excise this line, but keep the one below it specifically about 'node-'

  If you want to build the module as a Node package instead of for the browser, prefix the folder name with `node-`.

2. Copy the contents of `templates/module-template` into the new folder.

3. Edit `package.json`.

>   There is a few properties you should edit. They are: `name`, `description`, and `version`. Everything else is fine being left alone.

Reviewer 2: ```suggestion
  There are a few properties you should edit. They are: `name`, `description`, and `version`. Everything else is fine being left alone.
```

Reviewer 2: Any recommendations on sub-package versioning? I imagine it would probably be easiest to use [calver](https://calver.org/) for all of them, since "1.0" isn't really meaningful and they're entirely dependent on being in the Wikijump ecosystem.

Author: Lerna is in independent mode, which means packages get versioned arbitrarily depending on whether or not they have changed since they've last been published. I left versioning entirely alone in this because it's a bit of a can of worms. If you don't mind, I'd like to handle that in a separate PR when we start actually uploading packages.

Reviewer 2: If they're bumped automatically, what does that mean exactly? Does it just change the minor version every time or something?

This doesn't seem like an argument against calver per se, but I also don't really get what lerna does with versions so.

Author: To be honest, I haven't played with Lerna much. iirc, you give it `major | minor | patch` and it bumps? I have to look into it, and calver seems appealing to me.
